### PR TITLE
Make user keys endpoint specs pass

### PR DESCRIPTION
**Pull request by jrunning**
Depends on chef/chef-server#707.

- Add UserKeyEndpoint; handles `GET/DELETE/PUT /user_keys/USERNAME/keys/KEY_NAME`.
- Add UserKeysEndpoint; handles `GET/POST /user_keys/USERNAME`.
- RestBase
  - Add some docs.
  - `#json_response`, `#already_json_response`
    - Move `request_version` and `response_version` params into options hash.
    - Accept `:headers` option for additional headers to be set on response.
- RestObjectEndpoint
  - Add some docs.
  - Move some repeated logic into methods (`identity_key_value`, `is_rename?`).
  - `#patch_request_body`: Add `:except` option to allow skipping keys not wanted in output.
- ActorsEndpoint
  - `#post`: Store `public_key` in store under `user_keys/USERNAME/keys/default` instead of with user.
- ActorEndpoint
  - `#delete`: Delete user keys when user is deleted.
  - `#put`: Store `public_key` in store under `user_keys` as above.
  - `#populate_defaults`: Retrieve user's default `public_key` from store and merge with user response.
- OrganizationUserEndpoint
  - `#get`: Retrieve user's default `public_key` from store and merge with user response.
- PrincipalEndpont
  - `#get`
    - Retrieve user's default `public_key` from store and merge with user response.
    - Refactor complex nested `if`s.

### CODE REVIEW

**Comment by chef-supermarket** [comment]
Hi. I am an automated pull request bot named Curry. There are commits in this pull request whose authors are not yet authorized to contribute to Chef Software, Inc. projects or are using a non-GitHub verified email address. To become authorized to contribute, you will need to sign the Contributor License Agreement (CLA) as an individual or on behalf of your company. [You can read more on Chef's blog.](https://www.getchef.com/blog/2014/06/23/changes-to-the-contributor-license-agreement-process)

## GitHub Users Who Are Not Authorized To Contribute

The following GitHub users do not appear to have signed a CLA:

* @jrunning

[Please sign the CLA here.](https://supermarket.chef.io/become-a-contributor)

File: spec/run_oc_pedant.rb [hunk 1]
```diff
@@ -100,7 +100,8 @@ def start_cheffs_server(chef_repo_path)
     # are turned off" - @jkeiser
     #
     # ...but we're not there yet
-    '--skip-keys',
+    '--skip-client-keys',
+    '--skip-controls',
```
Comment (danielsdeleo) [inline spec/run_oc_pedant.rb]: Probably need to remove this after rebasing, since I think that's fixed now.

**Comment by danielsdeleo** [comment]
Can you give this a rebase so at least travis can run? Code looks good to me, :+1:

**Comment by chef-supermarket** [comment]
Hi. Your friendly Curry bot here. Just letting you know that all commit authors have become authorized to contribute. I have added the "Signed CLA" label to this issue so it can easily be found in the future.

**Comment by jrunning** [comment]
I rebased this, but it depends on `chef/chef-server#707` being merged, so it's failing on Travis right now.

**Comment by jrunning** [comment]
Actually, this is going to fail as long as we have this in the Gemfile on master:

```ruby
gem 'oc-chef-pedant', :github => 'chef/chef-server', :branch => "jk/authorization-tags"
```

**Comment by jrunning** [comment]
Closing this in favor of #196, which includes the commit from this PR.

**Pull request closed without merging**